skyjo_immutable.py

```diff
@@ -0,0 +1,513 @@
+import dataclasses
+import enum
+import random
+import typing
+from functools import cached_property
+from typing import Self
+
+import numpy as np
+
+NUM_CARD_TYPES = 15
+NUM_ROWS = 3
+NUM_COLUMNS = 4
+DECK_SIZE = 150
+DECK = (
+    (-2,) * 5
+    + (-1,) * 10
+    + (0,) * 15
+    + (1,) * 10
+    + (2,) * 10
+    + (3,) * 10
+    + (4,) * 10
+    + (5,) * 10
+    + (6,) * 10
+    + (7,) * 10
+    + (8,) * 10
+    + (9,) * 10
+    + (10,) * 10
+    + (11,) * 10
+    + (12,) * 10
+)
+
+
+class SkyjoActionType(enum.Enum):
+    DRAW = "DRAW"
+    PLACE_DRAWN = "PLACE_DRAWN"
+    PLACE_FROM_DISCARD = "PLACE_DISCARD"
+    DISCARD_AND_FLIP = "DISCARD_AND_FLIP"
+
+
+@dataclasses.dataclass(slots=True, frozen=True)
+class SkyjoAction:
+    action_type: SkyjoActionType = dataclasses.field()
+    row_idx: int | None = dataclasses.field(default=None)
+    col_idx: int | None = dataclasses.field(default=None)
+
+
+@dataclasses.dataclass(slots=True, frozen=True)
+class Hand:
+    is_cleared: np.array = dataclasses.field(
+        default_factory=lambda: np.zeros(NUM_ROWS * NUM_COLUMNS)
+    )
+    is_flipped: np.array = dataclasses.field(
```

Comment:
is it possible to just make this an array of `bool`? same for above? could you combine them into a single set of flags i.e. `state & 0b01` is cleared and `state & 0b10` is flipped?